### PR TITLE
type(bluebird): add `bluebird.allSettled`

### PR DESCRIPTION
**Pull request by bluelovers**
https://github.com/petkaantonov/bluebird/blob/750bd7f87fefaa0f918a6f0a25caec32ffdaddd8/docs/docs/changelog.md#370-2019-10-01

Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/bluebird/index.d.ts [hunk 1]
```diff
@@ -573,6 +574,13 @@ declare class Bluebird<R> implements PromiseLike<R>, Bluebird.Inspection<R> {
    */
   all(): Bluebird<never>;
 
+  static allSettled<T1, T2, T3, T4, T5>(values: [Resolvable<T1>, Resolvable<T2>, Resolvable<T3>, Resolvable<T4>, Resolvable<T5>]): Bluebird<ResultArrayToInspection<[T1, T2, T3, T4, T5]>>;
```
Comment (lhecker) [inline types/bluebird/index.d.ts:577]: Could you please move these definitions down to where all the other static definitions are?
I'd suggest moving them to 865, after the definitions for the static `all()` method.

**Review by lhecker — COMMENTED** [review]
In order for me to be comfortable approving this PR I'd like to ask you to extend the unit test file.
Let me know if you need help with that.

---

Some optional thoughts:
IMO it'd be better to remove `ResultArrayToInspection` and to simple inline the type definition into the static method definitions.
Currently each static type definition corresponds to one _single_ case of the ternary chain in `ResultArrayToInspection`. As such `ResultArrayToInspection` is IMO an unnecessary indirection at this point.
Of course I'm absolutely fine if you insist on keeping it.

**Review by lhecker — CHANGES_REQUESTED** [review]
(Forgot to mark this as "Request changes".)

File: types/bluebird/index.d.ts (hunk 1 above)
Comment (bluelovers) [reply types/bluebird/index.d.ts]: @lhecker done

**Comment by bluelovers** [comment]
> In order for me to be comfortable approving this PR I'd like to ask you to extend the unit test file.
> Let me know if you need help with that.

i don't know how add it

> IMO it'd be better to remove ResultArrayToInspection and to simple inline the type definition into the static method definitions.

hard to read when has so many `Bluebird.Inspection`

**Comment by lhecker** [comment]
## Regarding tests

[This file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f1008ba6f314c49abe8747e1da52ef684bce9d29/types/bluebird/bluebird-tests.ts) contains all tests for Bluebird. They aren't particularly beautiful but they work reasonably well.
[Here's the section for `Bluebird.all`.](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f1008ba6f314c49abe8747e1da52ef684bce9d29/types/bluebird/bluebird-tests.ts#L656-L676)

Please add another section below it which uses the new `settleAll` methods. The tests should ensure that the returned type matches what you expect. Nowadays we have `$ExpectType` declarations for tests which you may use for this. You can read about this [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f1008ba6f314c49abe8747e1da52ef684bce9d29/README.md#editing-tests-on-an-existing-package).

E.g. something like this (I didn't test this though):
```ts
// $ExpectType Bluebird<[Bluebird.Inspection<0>, Bluebird.Inspection<"1">, Bluebird.Inspection<2>]>
Bluebird.settleAll([0, "1", 2]);

// $ExpectType Bluebird<Bluebird.Inspection<number>[]>
Bluebird.settleAll(numArrProm);
```

## Regarding the new type

> hard to read when has so many `Bluebird.Inspection`

But now the `ResultArrayToInspection` type is hard to read instead.
This new type simply moves the "hard to read" part from one line to another.
It also doesn't reduce the amount of code either, but rather does the opposite.
Your PR adds 1489 characters of code. [If you remove the `ResultArrayToInspection` type](https://gist.github.com/lhecker/a0ef77cca8b74eac65fc40e5d4fd2504) it's only 982. A ~35% reduction!

A true "generic" solution would rather be a type which isn't bound to use `Bluebird.Inspection` for wrapping the array. And it could look similar to this:
```ts
type WrapArray<W<~>, T> =
    T extends [infer T1, infer T2, infer T3, infer T4, infer T5] ? [W<T1>, W<T2>, W<T3>, W<T4>, W<T5>] :
    T extends [infer T1, infer T2, infer T3, infer T4] ? [W<T1>, W<T2>, W<T3>, W<T4>] :
    T extends [infer T1, infer T2, infer T3] ? [W<T1>, W<T2>, W<T3>] :
    T extends [infer T1, infer T2] ? [W<T1>, W<T2>] :
    T extends [infer T1] ? [W<T1>] :
    T extends Iterable<infer T1> ? W<T1>[] :
    never;
```

It's relatively little code but is entirely generic and may be reused as often as you'd like.
But unfortunately you need higher kinded types for this and [TypeScript doesn't implement those yet](https://github.com/microsoft/TypeScript/issues/1213).

Considering the 25% reduction in code without `ResultArrayToInspection` I'd like to humbly ask you to inline the type please.

**Comment by bluelovers** [comment]
@lhecker done

**Review by lhecker — APPROVED** [review]
LGTM

**Comment by lhecker** [comment]
Ah you got 3 linter issues:
```
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/bluebird/index.d.ts:858:1
ERROR: 858:1   max-line-length  Exceeds maximum line length of 200
ERROR: 859:1   max-line-length  Exceeds maximum line length of 200
ERROR: 863:79  array-type       Array type using 'T[]' is forbidden for non-simple types. Use 'Array<T>' instead.
```

You can run `npm run lint bluebird` to verify your changes beforehand.
The `log4js` issues seem unrelated. As soon as they're fixed in master you might have to rebase your PR.

**Comment by bluelovers** [comment]
that error not from my edit lol

the exists code `static all` is use T[] too

**Comment by bluelovers** [comment]
@lhecker  other error not about bluebird now

**Comment by lhecker** [comment]
@bluelovers Apparently this issue is caused, because `log4js` was removed from this project more or less recently. The `log4js` project now has it's own typings.

The solution to this problem is to copy [this file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/karma/package.json) into the `v3` and `v4` subdirectories of karma.
Would you like me to open a separate PR to fix this issue or do you want to fix this yourself as part of this PR?

**Comment by lhecker** [comment]
I believe your rebase went wrong. Let me know if you need any help. :sweat_smile:

**Comment by bluelovers** [comment]
@lhecker lol, i think need help

**Comment by lhecker** [comment]
To fix the above do the following:
```sh
git reset --hard 2447b03
```

To fix the CI build do the following:
```sh
cp types/karma/package.json types/karma/v4/
cp types/karma/package.json types/karma/v3/
```

Finally push your changes:
```sh
git push -f
```

**Review by lhecker — APPROVED** [review]
I'm very sorry for the inconveniences you had, submitting this PR. Thanks for your hard work!

**Comment by bluelovers** [comment]
@lhecker thx help

**Review by devoto13 — APPROVED** [review]
Technically karma@3 depends on the log4js@3, but AFAICT there were no breaking changes between v3 and v4, so this should be pretty safe to do this change.

**Comment by sandersn** [comment]
@devoto good to know that in case of future problems, thanks for documenting it.

**Pull request merged**